core/src/main/scala/akka/kafka/internal/ActorConsumer.scala

```diff
@@ -0,0 +1,256 @@
+      .mapMaterializedValue(ActorControl(_))
+  }
+
+  class SubSourceActor[K, V](mainSource: ActorRef, tp: TopicPartition, kafka: ActorRef)
+      extends ActorPublisher[ConsumerRecord[K, V]] {
+    var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+    override def receive: Receive = normal
+
+    override def preStart(): Unit = {
+      super.preStart()
+      mainSource ! RegisterSubSource(tp)
+    }
+
+    @tailrec
+    private def pump(): Unit = {
+      if (buffer.hasNext && totalDemand > 0) {
+        onNext(buffer.next())
+        pump()
+      }
+    }
+
+    def normal: Receive = LoggingReceive {
+      case KafkaConsumerActor.Messages(msgs) =>
+        val newMessages = msgs.asInstanceOf[Iterator[ConsumerRecord[K, V]]]
+        buffer = buffer ++ newMessages
+        pump()
+      case Request(_) =>
+        pump()
+        if (!buffer.hasNext && totalDemand > 0) {
+          kafka ! KafkaConsumerActor.RequestMessages(Set(tp))
+        }
+      case Cancel => context.stop(self)
+      case ActorControl.Stop => onComplete()
+    }
+  }
+}
+
+class TopicPartitionSourceActor[K, V](settings: ConsumerSettings[K, V])
+    extends ActorPublisher[(TopicPartition, Source[ConsumerRecord[K, V], Control])] {
+  import TopicPartitionSourceActor._
+  var kafka: ActorRef = _
+  var buffer: immutable.Queue[TopicPartition] = immutable.Queue.empty
+  var subSources: Map[TopicPartition, ActorRef] = immutable.Map.empty
+
+  override def preStart(): Unit = {
+    super.preStart()
+    kafka = context.actorOf(Props(new KafkaConsumerActor(settings.createKafkaConsumer)))
+    kafka ! KafkaConsumerActor.Subscribe(settings.topics)
+  }
+
+  override def postStop(): Unit = {
+    subSources.values.foreach(context.stop)
+    context.stop(kafka)
+    super.postStop()
+  }
+
+  override def receive: Receive = LoggingReceive {
+    case RegisterSubSource(tp) =>
+      context.watch(sender)
+      subSources += (tp -> sender)
+    case KafkaConsumerActor.Assigned(tps) =>
+      buffer = buffer.enqueue(tps)
+      pump()
+    case KafkaConsumerActor.Revoked(tps) =>
+      tps.foreach { tp =>
+        subSources.get(tp) match {
+          case Some(ref) =>
+            context.unwatch(ref)
+            context.stop(ref)
+          case None =>
+            buffer = buffer.filter(x => tps.contains(x))
+        }
+      }
+    case Terminated(ref) =>
+      subSources.find(_._2 == ref) match {
+        case Some((tp, _)) =>
+          subSources -= tp
+          buffer :+= tp
+          pump()
+        case None =>
+      }
+    case Request(_) => pump()
+    case Cancel => context.stop(self)
+    case ActorControl.Stop =>
+      sender ! ActorControl.Stopped
+      onComplete()
+  }
+
+  def createSource(tp: TopicPartition): Source[ConsumerRecord[K, V], Control] = {
+    Source.actorPublisher(Props(new SubSourceActor(self, tp, kafka)))
```

Comment:
I think this should have a `SubSourceActor.props` to make sure that it doesn't close over something bad